Add  ignore_metadata_file option to read_parquet (pyarrow-dataset and fastparquet support only)

rjzamora opened the pull request:
May address #8030

This PR makes it possible to ignore a global _metadata file using the `ignore_metadata_file` argument:

```python
ddf = dd.read_parquet(
    path,
    engine="pyarrow-dataset",
    ignore_metadata_file=True,
)
```

~Note that I did not add support for fastparquet or pyarrow-legacy (they will both raise an error if `ignore_metadata_file=True`), only because it is much trickier to implement for those engines (and because the pyarrow-legacy engine will be deprecated soon anyway).~

**EDIT**: This PR now adds `ignore_metadata_file` support for both "pyarrow-datset" **and** "fastparquet".  The "pyarrow-legacy" case still doesn't make much sense because: (1) it is deprecated, and (2) the engine will still construct a proxy _metadata file anyway.

jrbourbeau reviewed (commented):
Thanks @rjzamora! Pardon my ignorance about our parquet engines, but I'm curious why supporting `ignore_metadata_file` is more involved for `fastparquet` (also cc @martindurant)

rjzamora commented:
>Thanks @rjzamora! Pardon my ignorance about our parquet engines, but I'm curious why supporting ignore_metadata_file is more involved for fastparquet

You are absolutely right to ask this, because I **should** be able to add support for fastparquet.  We will probably just need to use  some variation of `fs.find` to discover all the nested paths before passing to `fp.ParquetFile`.  I just noticed earlier that "fastparquet" was failing without the _metadata file for partitioned datasets, and I was not feeling particularly motivated to track down the root problem.  However, now that a larger `read_metadata` revision is in the cards, I am feeling a bit differently :)

jrbourbeau commented on dask/dataframe/io/parquet/arrow.py:
```diff
@@ -1030,7 +1045,7 @@ def _gather_metadata(
 
             meta_path = fs.sep.join([paths, "_metadata"])
-            if fs.exists(meta_path):
+            if fs.exists(meta_path) and not ignore_metadata_file:
```
What you have here is totally correct, but if we put the `ignore_metadata_file` condition first we might be able to avoid an unnecessary `fs.exists` call. This might not end up being important in practice, but I currently have S3 limits on the top of my mind : )

```suggestion
            if not ignore_metadata_file and fs.exists(meta_path):
```

jrbourbeau commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -2216,15 +2216,15 @@ def test_pathlib_path(tmpdir, engine):
 
 
-@PYARROW_MARK
+@PYARROW_LE_MARK
```
There are several changes to seemingly unrelated tests. I'm curious why these were needed

rjzamora replied:
Yes, good catch.  These test changes are mostly unrelated to logic added to this PR.  However, this PR does add a `dataset_kwargs` check to the pyarrow engine to avoid cases where we are silently ignoring user-privided options.  This small change triggered some CI failures that were mostly related to legacy-vs-dataset behavior.

jrbourbeau replied:
Hmm does that mean the `dataset=` is supported for the `pyarrow-legacy` engine? Naively I thought it was only supported when using the `pyarrow-dataset` engine

rjzamora replied:
That's right. The `dataset=` option was introduced before `pyarrow.dataset` existed. It was meant to mean "pass these kwargs to `ParquetDataset`".

rjzamora commented:
**Update**: Thanks for the reivew here @jrbourbeau ! I addressed your comments, and added "fastparquet" support here.  So, hopefully this PR is in good shape (once CI is passing).

martindurant commented on dask/dataframe/io/parquet/fastparquet.py:
```diff
@@ -146,13 +155,25 @@ def _determine_pf_parts(fs, paths, gather_statistics, **kwargs):
+        # This is a directory.
+        # Check if _metadata and/or _common_metadata files exists
+        base = paths[0]
+        _metadata_exists = fs.isfile(fs.sep.join([base, "_metadata"]))
```
If `ignore_metadata_file`, then we know we will have to list the whole directory anyway, so we can avoid the calls to `isfile` and even `isdir`, above.

rjzamora replied:
Good point - I just pushed a change to avoid the `isfile` calls.  However, it seems like we may need to move around the logic quite a bit to avoid the `isdir` call.  In fact, I'm nor really sure of an elegant way to avoid it.

martindurant reviewed (commented):
At a quick glance, the fp route looks OK, but first I wonder: should we explicitly disallow the combination of ignore_metadata_file with gather_statistics?

rjzamora commented:
>At a quick glance, the fp route looks OK, but first I wonder: should we explicitly disallow the combination of ignore_metadata_file with gather_statistics?

My vote is "no" - I don't think a willingness to ignore the shared _metadata file means the user should not be able to gather statistics. After #8058, there may even be cases where a parallel `read_metadata` with `ignore_metadata_file=True` will be quite fast.

jrbourbeau commented on dask/dataframe/io/parquet/core.py:
```diff
@@ -176,6 +177,8 @@ def read_parquet(
         statistics will only be gathered if True, because the footer of
         every file will be parsed (which is very slow on some systems).
+    ignore_metadata_file : bool, default False
+        Whether to ignore the global _metadata file (when one is present).
```
Nit

```suggestion
        Whether to ignore the global ``_metadata`` file (when one is present).
```

jrbourbeau commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -3580,3 +3587,32 @@ def test_custom_metadata(tmpdir, engine):
     assert "User-defined key/value" in str(e.value)
+
+
+@pytest.mark.parametrize("gather_statistics", [True, False])
```
Should we check `gather_statistics=None` too?

jrbourbeau commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -3580,3 +3587,32 @@ def test_custom_metadata(tmpdir, engine):
+        )
+
+        if gather_statistics is False:
+            # Should not have known divisions if gather_statistics=None
+            # Otherwise, we must have used a _metadata file
+            assert set(ddf2.divisions) == {None}
```
This seems fine, but I wonder if there's a more direct check we can have that `_metadata` wasn't used. For example, we could manually delete the `_metadata` file before calling `read_parquet`

martindurant replied:
Without the file, we would presumably get the same behaviour whether you have ignore-metadata true or false.

rjzamora replied:
I updated the test to do something pretty simple: We write two datasets (one with _metadata, and one without), and check that the result is the same with `ignore_metadata_file`.  How does that sound?

jrbourbeau replied:
I'll defer to you two on this test. I think I still need to wrap my head around how we handle gathering metadata and how that translates to division information on the resulting DataFrame. If you two are happy with this, then let's go with it

jorisvandenbossche replied:
Another option might be to create a dataset with a "wrong" `_metadata` file (eg only listing a subset of the files, or simply an invalid parquet file), and then reading it with `ignore_metadata_file`, you can be sure it was not used if you get the correct result.

martindurant replied:
That's the answer, thanks @jorisvandenbossche - replace the _metadata file with something invalid (perhaps empty?)

rjzamora replied:
That seems like a good approach, I pushed de44f6a (where I copied in a "wrong" _metadata file) and then b472c2b (where I wrote out an "invalid" file).  I'm going with the "invalid" approach for now, but happy to tweak things again.

jrbourbeau reviewed (commented):
Woo, thanks for adding `ignore_metadata_file` support to `fastparquet`!

jrbourbeau commented:
Just checking in, @rjzamora @martindurant is there anything else left to address here, or are we good to merge? Also cc @jorisvandenbossche for visibility

martindurant commented on dask/dataframe/io/parquet/fastparquet.py:
```diff
@@ -118,21 +118,30 @@ def _determine_pf_parts(fs, paths, gather_statistics, **kwargs):
-            if "_metadata" not in fns:
-                paths_use = paths
+            if _metadata_exists:
+                paths_use = fs.sep.join([base, "_metadata"])
```
I think I get confused here - don't we end up using _metadata even with ignore_metadata_file, if we asked for statistics? That's the inverse of my previous objection, but feels wrong this way around too, especially for gather_statistics=None

rjzamora replied:
>don't we end up using _metadata even with ignore_metadata_file, if we asked for statistics?

If `ignore_metadata_file=True`, we will not use the `_metadata` file.  However, we may and up gathering statistics manually.  We do not want `ignore_metadata_file` to be synonymous with `gather_statistics`.

To clarify further, I personally **don't** feel that the default behavior of `gather_statistics` should be `False` when the `_metadata` file is missing or ignored. If the user is not filtering, *and* is not setting and index, *and* is not aggregating files, then sure, default-False makes sense.

martindurant replied:
Oh OK never mind - you set `_metadata_exists = False` a few lines above (so it no longer means what it says it does! :) ). That was my confusion - I agree with you on the behaviour we want.

jrbourbeau commented:
Thanks for the test update @rjzamora -- I feel confident we're effectively testing that `_metadata` isn't read when `ignore_metadata_file=True`. There's a `test_partition_on` failure in CI, but from what I can tell it looks unrelated. Can you confirm?

rjzamora commented:
> There's a test_partition_on failure in CI, but from what I can tell it looks unrelated. Can you confirm?

Yikes.  I'd rather not ignore a failure in `test_parquet.py`.  I'll try to figure out what is going wrong.

martindurant commented:
>  I'll try to figure out what is going wrong.

Sounds like an odd one, we can pair to figure this out, if you like.

rjzamora commented:
@martindurant - Are you able to reproduce the failure locally? I'm not sure how any of the fastparquet-engine changes could be causing this.

martindurant commented:
Passes with dask main/fastparquet main
Passes with your branch/fastparquet main

Since it's only coming up on one build, could it be some combination of the random numbers, e.g., one of the specific sub-directories ends up empty?

martindurant commented:
I ran with `--count 30` and got no failures; but this is on mac.

jrbourbeau commented:
Just checking in, any luck reproducing? I'm inclined to merge if we're not able to and just keep an eye on `test_partition_on`

rjzamora commented:
>Just checking in, any luck reproducing? I'm inclined to merge if we're not able to and just keep an eye on test_partition_on

No luck reproducing on my end. We may have some problematic behavior to fix, but I suspect it is not being caused by this PR.

jrbourbeau commented:
Alright, thanks for your efforts on trying to reproduce @rjzamora @martindurant. I'll plan to merge this later in the evening if there aren't any objections
